### ISSUE TITLE
Boost 1.62 error

### ISSUE THREAD

**Issue by IanLee1521** [LLNL/spack#1903]
This one is relatively new... Somewhere between 8249108cc6ef82b954ed5b5eabaff88be42ee2b8..544fa3eddb2e0e3830439204888a58a9c0e186e2 (current HEAD of develop). I'm on OS X 10.11.6:

```
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.11.6
BuildVersion:	15G1004
```

```
$ ./bin/spack install boost
==> Installing boost
==> bzip2 is already installed in /Users/lee1001/projs/llnl/spack/opt/spack/darwin-elcapitan-x86_64/gcc-4.2.1/bzip2-1.0.6-gh6bktroppxbca4kxgmrjfezzrknb6he
==> zlib is already installed in /Users/lee1001/projs/llnl/spack/opt/spack/darwin-elcapitan-x86_64/gcc-4.2.1/zlib-1.2.8-x6uezf2iognzftplcz4tiesur3pfe6yx
==> Can not find version 1_62_0 in url_list
==> Already downloaded /Users/lee1001/projs/llnl/spack/var/spack/stage/boost-1.62.0-uiqanyaxjhf4iv4al7zyqnhqana2eo2z/boost_1_62_0.tar.bz2
==> Already staged boost-1.62.0-uiqanyaxjhf4iv4al7zyqnhqana2eo2z in /Users/lee1001/projs/llnl/spack/var/spack/stage/boost-1.62.0-uiqanyaxjhf4iv4al7zyqnhqana2eo2z
==> No patches needed for boost
==> Building boost
==> Error: Command exited with status 1:
'./b2' 'install' 'threading=multi' '-j' '8' 'variant=release' '-s' 'BZIP2_INCLUDE=/Users/lee1001/projs/llnl/spack/opt/spack/darwin-elcapitan-x86_64/gcc-4.2.1/bzip2-1.0.6-gh6bktroppxbca4kxgmrjfezzrknb6he/include' '-s' 'BZIP2_LIBPATH=/Users/lee1001/projs/llnl/spack/opt/spack/darwin-elcapitan-x86_64/gcc-4.2.1/bzip2-1.0.6-gh6bktroppxbca4kxgmrjfezzrknb6he/lib' '-s' 'ZLIB_INCLUDE=/Users/lee1001/projs/llnl/spack/opt/spack/darwin-elcapitan-x86_64/gcc-4.2.1/zlib-1.2.8-x6uezf2iognzftplcz4tiesur3pfe6yx/include' '-s' 'ZLIB_LIBPATH=/Users/lee1001/projs/llnl/spack/opt/spack/darwin-elcapitan-x86_64/gcc-4.2.1/zlib-1.2.8-x6uezf2iognzftplcz4tiesur3pfe6yx/lib' 'link=static,shared' '--layout=tagged' 'toolset=darwin'

See build log for details:
  /private/var/folders/vr/7lfgcgzd0wb4brywdhg86ygw00157w/T/lee1001/spack-stage/spack-stage-PZdvqg/boost_1_62_0/spack-build.out

/Users/lee1001/projs/llnl/spack/var/spack/repos/builtin/packages/boost/package.py:287, in install:
     226      def install(self, spec, prefix):
     227          # On Darwin, Boost expects the Darwin libtool. However, one of the
     228          # dependencies may have pulled in Spack's GNU libtool, and these two
     229          # are not compatible. We thus create a symlink to Darwin's libtool
     230          # and add it at the beginning of PATH.
     231          if sys.platform == 'darwin':
     232              newdir = os.path.abspath('darwin-libtool')
     233              mkdirp(newdir)
     234              force_symlink('/usr/bin/libtool', join_path(newdir, 'libtool'))
     235              env['PATH'] = newdir + ':' + env['PATH']
     236
     237          withLibs = list()
     238          for lib in Boost.all_libs:
     239              if "+{0}".format(lib) in spec:
     240                  withLibs.append(lib)
     241          if not withLibs:
     242              # if no libraries are specified for compilation, then you dont have
     243              # to configure/build anything, just copy over to the prefix
     244              # directory.
     245              src = join_path(self.stage.source_path, 'boost')
     246              mkdirp(join_path(prefix, 'include'))
     247              dst = join_path(prefix, 'include', 'boost')
     248              install_tree(src, dst)
     249              return
     250
     251          # Remove libraries that the release version does not support
     252          if not spec.satisfies('@1.54.0:'):
     253              withLibs.remove('log')
     254          if not spec.satisfies('@1.53.0:'):
     255              withLibs.remove('atomic')
     256          if not spec.satisfies('@1.48.0:'):
     257              withLibs.remove('locale')
     258          if not spec.satisfies('@1.47.0:'):
     259              withLibs.remove('chrono')
     260          if not spec.satisfies('@1.43.0:'):
     261              withLibs.remove('random')
     262          if '+graph' in spec and '+mpi' in spec:
     263              withLibs.remove('graph')
     264              withLibs.append('graph_parallel')
     265
     266          # to make Boost find the user-config.jam
     267          env['BOOST_BUILD_PATH'] = './'
     268
     269          bootstrap = Executable('./bootstrap.sh')
     270
     271          bootstrap_options = ['--prefix=%s' % prefix]
     272          self.determine_bootstrap_options(spec, withLibs, bootstrap_options)
     273
     274          bootstrap(*bootstrap_options)
     275
     276          # b2 used to be called bjam, before 1.47 (sigh)
     277          b2name = './b2' if spec.satisfies('@1.47:') else './bjam'
     278
     279          b2 = Executable(b2name)
     280          b2_options = ['-j', '%s' % make_jobs]
     281
     282          threadingOpts = self.determine_b2_options(spec, b2_options)
     283
     284          # In theory it could be done on one call but it fails on
     285          # Boost.MPI if the threading options are not separated.
     286          for threadingOpt in threadingOpts:
  >> 287              b2('install', 'threading=%s' % threadingOpt, *b2_options)
     288
     289          # The shared libraries are not installed correctly
     290          # on Darwin; correct this
     291          if (sys.platform == 'darwin') and ('+shared' in spec):
     292              fix_darwin_install_name(prefix.lib)
==> Error: Installation process had nonzero exit code : 256
```

Full log: [spack-build.txt](https://github.com/LLNL/spack/files/509426/spack-build.txt)

**Comment by alalazo** [comment]
Confirmed. Maybe the latest Boost deals with the comment above line 287 differently from what older versions did ?

**Comment by davydden** [comment]
also see it here with Ubuntu + GCC. A quick fix for now would be to set `1.61.0` as preferred until someone who knows more about boost can look at it.

**Comment by alalazo** [comment]
If somebody is going to touch boost, we should deal maybe with #1820 also

**Comment by davydden** [comment]
https://github.com/LLNL/spack/pull/1955 fixed this for me both on Ubuntu 16.04 and macOS Sierra. Can this issue be closed?
